From 8a1db7350d1037671747fbf135204c8976fb60fd Mon Sep 17 00:00:00 2001
From: Anthony Green <green@moxielogic.com>
Date: Wed, 28 Oct 2026 06:23:32 -0400
Subject: [PATCH] mips: fix O32 complex call and closure returns

Handle all integer-complex sizes in the O32 call path and return soft-float complex values through the ABI-defined GPR set. Restore hard-float closure results from the FP save area, preserve the argument flags, and pass the complex result size back to assembly explicitly.

Reported-by: Daniel Golle <daniel@makrotopia.org>
---
 src/mips/ffi.c | 50 ++++++++++++++++++++++++++++++++++++++++++++------
 src/mips/o32.S | 45 ++++++++++++++++++++++++++++++++++++++++-----
 2 files changed, 84 insertions(+), 11 deletions(-)

--- a/src/mips/ffi.c
+++ b/src/mips/ffi.c
@@ -402,7 +402,7 @@ static ffi_status ffi_prep_cif_machdep_i
    * does not have special handling for floating point args.
    */
 
-  if (cif->rtype->type != FFI_TYPE_STRUCT && cif->rtype->type != FFI_TYPE_COMPLEX && cif->abi == FFI_O32)
+  if (cif->rtype->type != FFI_TYPE_STRUCT && cif->abi == FFI_O32)
     {
       if (cif->nargs > 0 && cif->nargs == nfixedargs)
 	{
@@ -455,6 +455,24 @@ static ffi_status ffi_prep_cif_machdep_i
           cif->flags += FFI_TYPE_UINT64 << (FFI_FLAG_BITS * 2);
           break;
       
+        case FFI_TYPE_COMPLEX:
+          /* Complex values are returned in GPRs like integer complex
+             values of the same size.  */
+          cif->flags += FFI_TYPE_COMPLEX << (FFI_FLAG_BITS * 2);
+          switch (cif->rtype->elements[0]->type)
+            {
+            case FFI_TYPE_FLOAT:
+              cif->flags += FFI_TYPE_UINT32 << (FFI_FLAG_BITS * 4);
+              break;
+            case FFI_TYPE_DOUBLE:
+              cif->flags += FFI_TYPE_UINT64 << (FFI_FLAG_BITS * 4);
+              break;
+            default:
+              cif->flags += cif->rtype->elements[0]->type << (FFI_FLAG_BITS * 4);
+              break;
+            }
+          break;
+
         case FFI_TYPE_FLOAT:
         default:
           cif->flags += FFI_TYPE_INT << (FFI_FLAG_BITS * 2);
@@ -978,11 +996,19 @@ ffi_closure_mips_inner_O32 (ffi_cif *cif
       argn = 1;
       seen_int = 1;
     }
-  if ((cif->flags >> (FFI_FLAG_BITS * 2)) == FFI_TYPE_COMPLEX)
+  if (cif->rtype->type == FFI_TYPE_COMPLEX
+      && (cif->rtype->elements[0]->type == FFI_TYPE_FLOAT
+          || cif->rtype->elements[0]->type == FFI_TYPE_DOUBLE
+          || cif->rtype->elements[0]->type == FFI_TYPE_LONGDOUBLE))
     {
       rvalue = fpr;
-      argn = 1;
     }
+  /* With soft float, complex values are returned in up to four GPRs,
+     which does not fit into the space for v0 and v1.  ffi_closure_O32
+     does not use the space for the FP argument registers in that case,
+     so return the value there.  */
+  if (cif->abi == FFI_O32_SOFT_FLOAT && cif->rtype->type == FFI_TYPE_COMPLEX)
+    rvalue = fpr;
 
   i = 0;
   avn = cif->nargs;
@@ -1041,6 +1067,15 @@ ffi_closure_mips_inner_O32 (ffi_cif *cif
   /* Invoke the closure. */
   fun(cif, rvalue, avaluep, user_data);
 
+  if (cif->rtype->type == FFI_TYPE_COMPLEX && cif->rtype->size == 2)
+    {
+      /* A _Complex char is returned right-justified in v0.  */
+      UINT16 value;
+
+      memcpy (&value, rvalue, sizeof (value));
+      *(UINT32 *) rvalue = value;
+    }
+
   if (cif->abi == FFI_O32_SOFT_FLOAT)
     {
       switch (cif->rtype->type)
@@ -1049,15 +1084,18 @@ ffi_closure_mips_inner_O32 (ffi_cif *cif
           return FFI_TYPE_INT;
         case FFI_TYPE_DOUBLE:
           return FFI_TYPE_UINT64;
+        case FFI_TYPE_COMPLEX:
+          return (FFI_TYPE_COMPLEX
+                  | (cif->rtype->size << (FFI_FLAG_BITS * 2)));
         default:
           return cif->rtype->type;
         }
     }
   else
     {
-      if (cif->rtype->type == FFI_TYPE_COMPLEX) {
-          __asm__ volatile ("move $v1, %0" : : "r"(cif->rtype->size));
-      }
+      if (cif->rtype->type == FFI_TYPE_COMPLEX)
+        return (FFI_TYPE_COMPLEX
+                | (cif->rtype->size << (FFI_FLAG_BITS * 2)));
       return cif->rtype->type;
     }
 }
--- a/src/mips/o32.S
+++ b/src/mips/o32.S
@@ -224,8 +224,25 @@ retcomplex:
 	b	epilogue
 
 5:
+	# Integer complex values: a _Complex char is returned right-justified
+	# in v0, a _Complex short in v0, a _Complex int in v0/v1 and a
+	# _Complex long long in v0/v1/a0/a1.
+	slti	t3, t1, FFI_TYPE_UINT16
+	bnez	t3, 6f
+	slti	t3, t1, FFI_TYPE_UINT32
+	bnez	t3, 7f
 	REG_S	v1, 4(t0)
 	REG_S	v0, 0(t0)
+	slti	t3, t1, FFI_TYPE_UINT64
+	bnez	t3, epilogue
+	REG_S	a1, 12(t0)
+	REG_S	a0, 8(t0)
+	b	epilogue
+6:
+	sh	v0, 0(t0)
+	b	epilogue
+7:
+	REG_S	v0, 0(t0)
 	b	epilogue
 
 noretval:	
@@ -408,11 +425,13 @@ $do_closure:
 
 	# Load the return value into the appropriate register.
 	move	$8, $2
+	SRL	$3, $8, (FFI_FLAG_BITS * 2)
+	and	$8, ((1 << (FFI_FLAG_BITS * 2)) - 1)
 	li	$9, FFI_TYPE_VOID
 	beq	$8, $9, closure_done
 
 	li	$13, 1		# FFI_O32
-	bne	$16, $13, 1f	# Skip fp restore if FFI_O32_SOFT_FLOAT
+	bne	$16, $13, 2f	# Skip fp restore if FFI_O32_SOFT_FLOAT
 
 #ifndef __mips_soft_float
 	li	$9, FFI_TYPE_FLOAT
@@ -427,15 +446,31 @@ $do_closure:
 	bne     $8, $9, 1f
 
 	li      $9, 8
-	l.s	$f0, V0_OFF2($fp)
-	l.s	$f2, V1_OFF2($fp)
+	REG_L	$2, FA_0_0_OFF2($fp)
+	mtc1	$2, $f0
+	REG_L	$2, FA_0_1_OFF2($fp)
+	mtc1	$2, $f2
 	beq     $3, $9, closure_done
 
 	li      $9, 16
-	l.d	$f0, V0_OFF2($fp)
-	l.d	$f2, (V0_OFF2+8)($fp)
+	l.d	$f0, FA_0_0_OFF2($fp)
+	l.d	$f2, FA_1_0_OFF2($fp)
 	beq     $3, $9, closure_done
 #endif
+	b	1f
+
+2:
+	# FFI_O32_SOFT_FLOAT returns complex values in up to four GPRs.
+	# ffi_closure_mips_inner_O32 stored them in the space for the
+	# FP argument registers.
+	li	$9, FFI_TYPE_COMPLEX
+	bne	$8, $9, 1f
+	REG_L	$2, FA_0_0_OFF2($fp)
+	REG_L	$3, FA_0_1_OFF2($fp)
+	REG_L	$4, FA_1_0_OFF2($fp)
+	REG_L	$5, FA_1_1_OFF2($fp)
+	b	closure_done
+
 1:	
 	REG_L	$3, V1_OFF2($fp)
 	REG_L	$2, V0_OFF2($fp)
